File: src/lib/missions/handlers/ACSAttackMission.ts
/**
 * ACSAttackMission Handler
 *
 * Handles ACS (Alliance Combat System) attack missions where multiple fleets
 * attack together in a coordinated assault.
 *
 * Key Differences from Regular Attack:
 * - Waits for all participating fleets to arrive
 * - Combines all attacking fleets into one battle
 * - Distributes loot based on cargo capacity
 * - All participants receive battle reports
 */

import type { MissionType } from '@/types/database'
import { BaseMission } from '../BaseMission'
import {
  MissionContext,
  MissionArrivalResult,
  MissionReturnResult,
  Resources,
  ShipCounts,
  emptyResources,
  emptyShipCounts,
  SHIP_KEYS,
  hasResources,
} from '../types'
import { createACSService } from '@/lib/acs'
import type { ACSParticipantDB } from '@/types/acs'

export class ACSAttackMission extends BaseMission {
  readonly missionType: MissionType = 'acs_attack'
  readonly hasReturn: boolean = true
  readonly name: string = 'ACS Attack'

  /**
   * Process ACS attack mission arrival
   *
   * When a fleet arrives at the ACS target:
   * 1. Mark participant as arrived
   * 2. Check if all participants have arrived
   * 3. If all arrived, execute combined battle via ACSService
   * 4. If not all arrived, wait (fleet stays at target)
   */
  async processArrival(context: MissionContext): Promise<MissionArrivalResult> {
    const { mission, targetPlanet, attackerResearch } = context

    // Find the ACS operation this mission belongs to
    const { data: participant } = await this.supabase
      .from('acs_participants')
      .select('*, acs_operations(*)')
      .eq('fleet_mission_id', mission.id)
      .single()

    if (!participant) {
      // Not part of an ACS operation - shouldn't happen, but handle gracefully
      const ships = this.getShips(mission)
      const resources = this.getResources(mission)

      const message = this.createMessage(
        mission.user_id,
        'battle',
        'ACS Mission Error',
        'This fleet was not properly linked to an ACS operation. Returning home.'
      )

      return this.successArrival(true, resources, ships, [message], [])
    }

    const operation = participant.acs_operations

    // Mark this participant as arrived
    await this.supabase
      .from('acs_participants')
      .update({
        status: 'arrived',
        arrival_time: new Date().toISOString(),
      })
      .eq('id', participant.id)

    // Check if all participants have arrived
    const { data: allParticipants } = await this.supabase
      .from('acs_participants')
      .select('*')
      .eq('acs_operation_id', operation.id)
      .in('status', ['confirmed', 'en_route', 'arrived'])

    const arrivedCount = allParticipants?.filter(p => p.status === 'arrived').length || 0
    const totalCount = allParticipants?.length || 0

    if (arrivedCount < totalCount) {
      // Not all fleets have arrived yet
      // The fleet waits at the target (no immediate return)
      // The mission processor will check again later
      const message = this.createMessage(
        mission.user_id,
        'battle',
        'ACS Fleet Waiting',
        `Your fleet has arrived at the target coordinates and is waiting for ${totalCount - arrivedCount} more fleet(s) to arrive.`
      )

      // Don't return yet - fleet waits
      return this.successArrival(false, emptyResources(), emptyShipCounts(), [message], [])
    }

    // All fleets have arrived - execute the ACS battle
    const acsService = createACSService(this.supabase)

    // Update operation status to in_progress
    await this.supabase
      .from('acs_operations')
      .update({ status: 'in_progress' })
      .eq('id', operation.id)

    // Execute the combined battle
    const battleResult = await acsService.executeACSBattle(operation.id)

    if (!battleResult.success) {
      const ships = this.getShips(mission)
      const resources = this.getResources(mission)

      const message = this.createMessage(
        mission.user_id,
        'battle',
        'ACS Battle Error',
        `An error occurred during the ACS battle: ${battleResult.error}`
      )

      return this.successArrival(true, resources, ships, [message], [])
    }

    // Find this participant's results
    const myResult = battleResult.result!.participant_results.find(
      r => r.user_id === mission.user_id
    )

    if (!myResult) {
      const ships = this.getShips(mission)
      const resources = this.getResources(mission)
      return this.successArrival(true, resources, ships, [], [])
    }

    // Create battle message for this participant
    const targetCoords = `[${operation.target_galaxy}:${operation.target_system}:${operation.target_position}]`
    const winnerText = battleResult.result!.winner === 'attacker'
      ? 'The attackers have won the battle!'
      : battleResult.result!.winner === 'defender'
        ? 'The defender has won the battle!'
        : 'The battle ended in a draw.'

    const lossesText = this.formatShipLosses(myResult.ships_lost)
    const lootText = `Metal: ${myResult.loot_share.metal.toLocaleString()}\nCrystal: ${myResult.loot_share.crystal.toLocaleString()}\nDeuterium: ${myResult.loot_share.deuterium.toLocaleString()}`

    const message = this.createMessage(
      mission.user_id,
      'battle',
      `ACS Battle Report - ${targetCoords}`,
      `${winnerText}\n\n` +
        `Total rounds: ${battleResult.result!.total_rounds}\n` +
        `Total participants: ${totalCount}\n\n` +
        `Your losses:\n${lossesText || 'None'}\n\n` +
        `Your loot share:\n${lootText}\n\n` +
        `Debris field created:\n` +
        `Metal: ${battleResult.result!.debris.metal.toLocaleString()}\n` +
        `Crystal: ${battleResult.result!.debris.crystal.toLocaleString()}\n\n` +
        (battleResult.result!.moon_created
          ? `A moon has been created! (${battleResult.result!.moon_chance}% chance)`
          : battleResult.result!.moon_chance > 0
            ? `Moon chance: ${battleResult.result!.moon_chance}% (no moon created)`
            : '')
    )

    // Calculate return ships and resources
    const returnShips = this.convertToShipCounts(myResult.ships_remaining)
    const originalResources = this.getResources(mission)
    const returnResources: Resources = {
      metal: originalResources.metal + myResult.loot_share.metal,
      crystal: originalResources.crystal + myResult.loot_share.crystal,
      deuterium: originalResources.deuterium + myResult.loot_share.deuterium,
    }

    return this.successArrival(true, returnResources, returnShips, [message], [])
  }

  /**
   * Process ACS attack mission return
   */
  async processReturn(context: MissionContext): Promise<MissionReturnResult> {
    const { mission, originPlanet } = context

    if (!originPlanet) {
      return this.errorReturn('Origin planet not found')
    }

    const ships = this.getShips(mission)
    const resources = this.getResources(mission)

    // Add ships back to origin
    const addShipsResult = await this.addShipsToPlanet(originPlanet.id, ships)
    if (!addShipsResult.success) {
      return this.errorReturn(`Failed to return ships: ${addShipsResult.error}`)
    }

    // Add resources (including loot share)
    if (hasResources(resources)) {
      const addResourcesResult = await this.addResourcesToPlanet(originPlanet.id, resources)
      if (!addResourcesResult.success) {
        return this.errorReturn(`Failed to return resources: ${addResourcesResult.error}`)
      }
    }

    // Update participant status
    await this.supabase
      .from('acs_participants')
      .update({ status: 'returned' })
      .eq('fleet_mission_id', mission.id)

    const message = this.createMessage(
      mission.user_id,
      'battle',
      'ACS Fleet Returned',
      `Your fleet has returned from the ACS attack mission.${
        hasResources(resources)
          ? `\n\nResources carried:\nMetal: ${resources.metal.toLocaleString()}\nCrystal: ${resources.crystal.toLocaleString()}\nDeuterium: ${resources.deuterium.toLocaleString()}`
          : ''
      }`
    )

    return this.successReturn([message], [])
  }

  // ============================================================================
  // HELPER METHODS
  // ============================================================================

  private convertToShipCounts(ships: Record<string, number>): ShipCounts {
    const result = emptyShipCounts()
    for (const key of SHIP_KEYS) {
      result[key] = ships[key] || 0
    }
    return result
  }

  private formatShipLosses(losses: Record<string, number>): string {
    const lines: string[] = []
    for (const [key, count] of Object.entries(losses)) {
      if (count > 0) {
        const name = key.replace(/_/g, ' ').replace(/\b\w/g, l => l.toUpperCase())
        lines.push(`${name}: ${count.toLocaleString()}`)
      }
    }
    return lines.join('\n')
  }
}

File: unity_client/Assets/Scripts/SharedLib/Missions/Handlers/ACSAttackMission.cs
// =============================================================================
// ACSAttackMission.cs — Allied Combat System attack mission handler
// Port of: src/lib/missions/handlers/ACSAttackMission.ts
// Namespace: OGameX.SharedLib.Missions.Handlers
// =============================================================================

using System;
using System.Collections.Generic;
using System.Linq;

namespace OGameX.SharedLib.Missions.Handlers
{
    /// <summary>
    /// Handles ACS (Alliance/Allied Combat System) attack missions.
    /// 
    /// ACS attacks coordinate multiple fleets from different players
    /// to attack a single target simultaneously. Flow:
    /// 1. Each ACS participant sends a fleet with mission type ACSAttack
    /// 2. All fleets share an ACS group ID
    /// 3. When the last fleet arrives (or timeout), battle executes
    /// 4. Battle includes all participating attacker fleets vs. defender
    /// 5. Loot is distributed among attackers proportionally
    /// 6. Each fleet returns independently with their share
    /// 
    /// Note: The ACS coordination logic (waiting, grouping) is handled
    /// at the data layer. This handler processes the actual combat once
    /// all participants have arrived.
    /// </summary>
    public class ACSAttackMission : BaseMission
    {
        public override MissionType HandlerMissionType => MissionType.ACSAttack;
        public override bool HasReturn => true;
        public override string Name => "ACS Attack";

        // =====================================================================
        // ACS GROUP TYPES
        // =====================================================================

        /// <summary>Represents a participating fleet in an ACS attack.</summary>
        public class ACSParticipant
        {
            public string UserId { get; set; } = "";
            public FleetMission Mission { get; set; } = new();
            public ShipCounts Ships { get; set; } = new();
            public UserResearch? Research { get; set; }
        }

        // =====================================================================
        // ARRIVAL PROCESSING
        // =====================================================================

        public override MissionArrivalResult ProcessArrival(MissionContext context)
        {
            var mission = context.Mission;
            var targetPlanet = context.TargetPlanet;

            if (targetPlanet == null)
                return ErrorArrival("Target planet not found");

            var ships = GetShips(mission);
            var originCoords = FormatCoords(
                mission.OriginGalaxy, mission.OriginSystem, mission.OriginPosition);
            var targetCoords = FormatCoords(
                mission.DestinationGalaxy, mission.DestinationSystem, mission.DestinationPosition);

            // In a full implementation, we check if all ACS members have arrived
            // For the shared library, we process this fleet's contribution
            var isACSReady = CheckACSGroupReady(context);

            if (!isACSReady)
            {
                // Not all participants arrived yet — mark as waiting
                var waitMsg = CreateMessage(
                    mission.UserId,
                    MessageType.Battle,
                    "ACS Attack - Waiting",
                    $"Your fleet has arrived at {targetCoords} and is waiting for allied fleets."
                );

                return SuccessArrival(
                    shouldReturn: false,
                    messages: new List<MissionMessage> { waitMsg }
                );
            }

            // All participants ready — execute combined attack
            var participants = GetACSParticipants(context);
            var combinedResult = ExecuteACSBattle(participants, targetPlanet, context);

            // Distribute loot proportionally among participants
            var lootShare = DistributeLoot(
                combinedResult.Loot, mission, participants);

            // Build messages
            var messages = new List<MissionMessage>();

            // Report for this participant
            messages.Add(CreateMessage(
                mission.UserId,
                MessageType.Battle,
                "ACS Attack Report",
                $"ACS attack at {targetCoords}\n\n" +
                $"Participants: {participants.Count}\n" +
                $"Result: {FormatBattleWinner(combinedResult.Winner)}\n\n" +
                $"Your loot share:\n{FormatResources(lootShare)}\n\n" +
                $"Total rounds: {combinedResult.TotalRounds}"
            ));

            // Report for defender
            if (targetPlanet.UserId != mission.UserId)
            {
                messages.Add(CreateMessage(
                    targetPlanet.UserId,
                    MessageType.Battle,
                    "ACS Attack Report (Defense)",
                    $"An ACS attack from {participants.Count} fleets hit your planet at {targetCoords}.\n\n" +
                    $"Result: {FormatBattleWinner(combinedResult.Winner)}\n\n" +
                    $"Your losses:\n{FormatShipLosses(combinedResult.DefenderShipLosses)}"
                ));
            }

            return SuccessArrival(
                shouldReturn: true,
                returnResources: lootShare,
                returnShips: GetSurvivingShips(ships, combinedResult),
                messages: messages
            );
        }

        // =====================================================================
        // RETURN PROCESSING
        // =====================================================================

        public override MissionReturnResult ProcessReturn(MissionContext context)
        {
            var mission = context.Mission;
            var resources = GetResources(mission);

            var originCoords = FormatCoords(
                mission.DestinationGalaxy, mission.DestinationSystem, mission.DestinationPosition);

            string messageBody = $"Your ACS attack fleet has returned to {originCoords}";
            if (resources.HasResources())
                messageBody += $" carrying loot:\n{FormatResources(resources)}";
            else
                messageBody += ".";

            var message = CreateMessage(
                mission.UserId,
                MessageType.Battle,
                "ACS Fleet Returned",
                messageBody
            );

            return SuccessReturn(new List<MissionMessage> { message });
        }

        // =====================================================================
        // ACS GROUP COORDINATION (Virtual hooks)
        // =====================================================================

        /// <summary>
        /// Check if all ACS group participants have arrived.
        /// Override in game layer with actual group tracking.
        /// Default: always ready (single-fleet).
        /// </summary>
        protected virtual bool CheckACSGroupReady(MissionContext context)
        {
            return true;
        }

        /// <summary>
        /// Get all participant fleets for this ACS group.
        /// Override in game layer to fetch grouped fleets.
        /// Default: just the current mission.
        /// </summary>
        protected virtual List<ACSParticipant> GetACSParticipants(MissionContext context)
        {
            return new List<ACSParticipant>
            {
                new ACSParticipant
                {
                    UserId = context.Mission.UserId,
                    Mission = context.Mission,
                    Ships = GetShips(context.Mission),
                    Research = context.AttackerResearch,
                }
            };
        }

        /// <summary>
        /// Execute the combined ACS battle.
        /// Override with real battle engine. Default: attacker wins.
        /// </summary>
        protected virtual MissionBattleResult ExecuteACSBattle(
            List<ACSParticipant> attackers,
            PlanetData defender,
            MissionContext context)
        {
            // Default stub — merge all attacker ships and simulate
            var combined = ShipCounts.Empty();
            foreach (var p in attackers)
                combined.AddAll(p.Ships);

            return new MissionBattleResult
            {
                Winner = BattleWinner.Attacker,
                TotalRounds = 1,
                AttackerRemainingShips = combined,
                DefenderRemainingShips = ShipCounts.Empty(),
                AttackerShipLosses = ShipCounts.Empty(),
                DefenderShipLosses = defender.Ships.Clone(),
                Loot = new Resources(
                    Math.Floor(defender.Metal * 0.5),
                    Math.Floor(defender.Crystal * 0.5),
                    Math.Floor(defender.Deuterium * 0.5)
                ),
            };
        }

        // =====================================================================
        // LOOT DISTRIBUTION
        // =====================================================================

        /// <summary>
        /// Distribute loot proportionally based on each participant's fleet value.
        /// </summary>
        protected Resources DistributeLoot(
            Resources totalLoot,
            FleetMission thisFleet,
            List<ACSParticipant> participants)
        {
            if (participants.Count <= 1)
                return totalLoot;

            // Calculate fleet "power" for each participant (simple ship count)
            int totalPower = participants.Sum(p => p.Ships.GetTotal());
            if (totalPower == 0) totalPower = 1;

            int thisPower = GetShips(thisFleet).GetTotal();
            double share = (double)thisPower / totalPower;

            return new Resources(
                Math.Floor(totalLoot.Metal * share),
                Math.Floor(totalLoot.Crystal * share),
                Math.Floor(totalLoot.Deuterium * share)
            );
        }

        // =====================================================================
        // HELPERS
        // =====================================================================

        /// <summary>Get surviving ships for this fleet from combined result.</summary>
        private ShipCounts GetSurvivingShips(
            ShipCounts originalShips,
            MissionBattleResult result)
        {
            // Simplified: proportional survival based on total losses
            // Real impl should track per-fleet losses
            var remaining = originalShips.Clone();
            int totalOriginal = originalShips.GetTotal();
            if (totalOriginal == 0) return remaining;

            int totalRemaining = result.AttackerRemainingShips.GetTotal();
            int totalSent = 0;
            foreach (var ship in ShipCounts.AllTypes)
                totalSent += result.AttackerRemainingShips.Get(ship) + result.AttackerShipLosses.Get(ship);

            if (totalSent == 0) return remaining;

            double survivalRate = (double)totalRemaining / totalSent;
            foreach (var t in ShipCounts.AllTypes)
            {
                int original = originalShips.Get(t);
                remaining.Set(t, (int)Math.Floor(original * survivalRate));
            }

            return remaining;
        }

        private string FormatBattleWinner(BattleWinner winner) => winner switch
        {
            BattleWinner.Attacker => "Attacker wins!",
            BattleWinner.Defender => "Defender wins!",
            BattleWinner.Draw => "Draw",
            _ => "Unknown",
        };
    }
}
